aws/signer/v4/middleware.go

```diff
@@ -0,0 +1,176 @@
+package v4
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+
+	v4Internal "github.com/aws/aws-sdk-go-v2/aws/signer/internal/v4"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyHTTP "github.com/awslabs/smithy-go/transport/http"
+)
+
+// HashComputationError indicates an error occurred while computing the signing hash
+type HashComputationError struct {
+	Err error
+}
+
+// Error is the error message
+func (e *HashComputationError) Error() string {
+	return fmt.Sprintf("failed to compute payload hash: %v", e.Err)
+}
+
+// Unwrap returns the underlying error if one is set
+func (e *HashComputationError) Unwrap() error {
+	return e.Err
+}
+
+// SigningError indicates an error condition occurred while performing SigV4 signing
+type SigningError struct {
+	Err error
+}
+
+func (e *SigningError) Error() string {
+	return fmt.Sprintf("failed to sign request: %v", e.Err)
+}
+
+// Unwrap returns the underlying error cause
+func (e *SigningError) Unwrap() error {
+	return e.Err
+}
+
+// UnsignedPayloadMiddleware sets the SigV4 request payload hash to unsigned
+type UnsignedPayloadMiddleware struct{}
+
+// ID returns the UnsignedPayloadMiddleware identifier
+func (m *UnsignedPayloadMiddleware) ID() string {
+	return "SigV4UnsignedPayloadMiddleware"
+}
+
+// HandleFinalize sets the payload hash to be an unsigned payload
+func (m *UnsignedPayloadMiddleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (
+	out middleware.FinalizeOutput, metadata middleware.Metadata, err error,
+) {
+	ctx = SetPayloadHash(ctx, v4Internal.UnsignedPayload)
+	return next.HandleFinalize(ctx, in)
+}
+
+// ComputePayloadSHA256Middleware computes sha256 payload hash to sign
+type ComputePayloadSHA256Middleware struct{}
+
+// ID is the middleware name
+func (m *ComputePayloadSHA256Middleware) ID() string {
+	return "ComputePayloadSHA256Middleware"
+}
+
+// HandleFinalize compute the payload hash for the request payload
+func (m *ComputePayloadSHA256Middleware) HandleFinalize(ctx context.Context, in middleware.FinalizeInput, next middleware.FinalizeHandler) (
+	out middleware.FinalizeOutput, metadata middleware.Metadata, err error,
+) {
+	req, ok := in.Request.(*smithyHTTP.Request)
+	if !ok {
+		return middleware.FinalizeOutput{}, metadata, &HashComputationError{Err: fmt.Errorf("unexpected request middleware type %T", in.Request)}
+	}
+
+	body, ok := req.Stream.(io.ReadSeeker)
+	if !ok {
+		return middleware.FinalizeOutput{}, metadata, &HashComputationError{Err: fmt.Errorf("stream does not implement io.Seeker, got %T", req.Stream)}
+	}
+
+	_, err = body.Seek(0, io.SeekStart)
+	if err != nil {
+		return middleware.FinalizeOutput{}, metadata, &HashComputationError{Err: fmt.Errorf("failed to seek body to start, %w", err)}
+	}
```

Comment:
So we want to be careful about seeking the payload. Seeking to 0 vs the current position the stream was received at could cause unexpected behavior. Its an odd situation but should the SDK attempt to reseek back to the begin of a reader if the user gave it to the input parameter at some other offset. Good example of this is I want to upload only the trailing portion of a file.

this is the reason the current seeking behavior in the SDK keeps track of initial offset. Might make sense to put all of the seeking behavior into a helper since it will be reused by retry. Maybe as apart of smithyhttp.Request?